### ISSUE TITLE
Using instanced (and custom) nodes in the layout DSL

### ISSUE THREAD

**Issue by jordaorodolfo** [edvin/tornadofx#738]
Hello! I did not find anything that closely matches this issue, thus posted a new one. Also, sorry if this is somehow implied in the guide, for I did not grasp if it is.
The thing is that when a node (or view, or fragment etc...) is instanced before the root layout DSL of a View or Fragment, say:

    val statusCheckComboBox = CheckComboBox(possibleStatus)

    override val root = form {
        fieldset("Cool  label", labelPosition = Orientation.VERTICAL) {
            field("Cooler") { statusCheckComboBox }
        }
    }

where the `CheckComboBox` is the one from ControlsFX and `possibleStatus` is a simple observable list; the rendered layout in the running program does not contain the CheckComboBox, in fact, there is just empty space there.
Now, from my perspective, when there a this pre-instanced node that must be used in the root node, the usual way, extracted from the guide, is to do:

    root += extraNode1
    root += extraNode2

And honestly, this can also be achieved in the layout Tree if I would call the `field("Cooler")` some variable, say, `coolField` and do at the init:

    coolField += statusCheckComboBox

And that works perfectly, being exactly my workaround for the moment. My complain is that this can get a bit cumbersome the more deeply nested the instanced nodes must be, and add uncesessary references to nested nodes in what, for me, is a wonderful way to declare the layout.

For the record, I've got no specific knowledge of TornadoFX internals, which means that I do not know if the builders approach make this "instanced injection" impossible. I think it is just valid to point out this, Admittedly, boring issue.

Thanks.

**Comment by tieskedh** [comment]
I believe you can use `this`

```
val statusCheckComboBox = CheckComboBox(possibleStatus)

override val root = form {
    fieldset("Cool  label", labelPosition = Orientation.VERTICAL) {
        field("Cooler") { 
            this += statusCheckComboBox 
        }
    }
}
```

**Comment by edvin** [comment]
Yeah, @tieskedh's suggestion is the way to go for none-core JavaFX Components you create often. In fact, that's why we added the tornadofx-controlsfx builders project, to provide those builders for ControlsFX components. You might want to check it out and maybe even contribute some builders if you use components that don't have builders yet :)

If you don't want to create builders, you should take another look at your design - I can't see a reason for creating those ControlsFX elements before you construct the UI - why do you feel you need to do that? In the example above you could just do:

```kotlin
field("Cooler") {
    this += CheckComboBox(possibleStatus)
}
```

For existing View instances, available in the same Scope, you could just do `add<MyOtherView>()` instead.

If you feel you need to instantiate Nodes in different places, I think you should take another look at your design :)

**Comment by jordaorodolfo** [comment]
Thanks @tieskedh  and @edvin , the approaches you provided work flawlessly!

Indeed, my design choices do not require me to build anything before the UI, necessarily. I asked this question exactly because I did want to mangle UI DSL and logic-heavy parts and did not know how to solve it before ;). Having said that, this pattern may have arisen for me because I've been using fragments to make reusable self-contained components, in a manner that resembles somawhat both android and vue.js approaches to UI and logic encapsulation. Then, some time ago when I first tried to instance a fragment inside a view, it did not show up to my surprise!

Now, I think I've hit a sweet spot to make stuff work. You see, I am still not making use of ViewModels or any other cool facility of ToFX because the data-centric sections of the code is directly shared with an android app (both are internal software of a small company), and I'd prefer it remain as pure kotlin as possible. Hence, some of my early attempts at android may have biased me towards ToFX.

Anyway, I think this can be deemed solved. Thanks a lot you two for the hints and as soon as time pops up, I'll try to make builders for my custom components (and share them if they are from ControlsFX). I just would like to suggest that this approach be incorporated, even if faintly, in the guide; mainly at the 3. Components section. It could help a lot newbies like me with different backgrounds to save some headache later!

Best regards.
